include/dali/imgcodec/image_source.h

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_IMGCODEC_IMAGE_SOURCE_H_
+#define DALI_IMGCODEC_IMAGE_SOURCE_H_
+
+#include <memory>
+#include <string>
+#include <utility>
+#include "dali/core/api_helper.h"
+#include "dali/core/int_literals.h"
+#include "dali/core/stream.h"
+
+namespace dali {
+namespace imgcodec {
+
+enum class InputKind : int {
+  None = 0,
+  // abstract stream interface that reads data from a custom file-like source
+  Stream = 1,
+  // bitstream loaded into host memory
+  HostMemory = 2,
+  // bitstream loaded into device memory
+  DeviceMemory = 4,
+  // file name
+  Filename = 8,
+};
+
+constexpr InputKind operator|(InputKind a, InputKind b) {
+  return static_cast<InputKind>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+constexpr InputKind operator&(InputKind a, InputKind b) {
+  return static_cast<InputKind>(static_cast<int>(a) & static_cast<int>(b));
+}
+
+constexpr bool operator!(InputKind k) {
+  return k == InputKind::None;
+}
+
+/**
+ * @brief A source of data from image parsers and codecs
+ */
+class DLL_PUBLIC ImageSource {
```

Comment:
Can you explain a bit how `ImageSource` will be created? By the readers? Or output from the readers will be wrapped in the decoder?

Reply:
At least initially the decoder will wrap the memory into an `ImageSource` by calling `FromHostMem`. In case of fused reader-decoders, we will likely employ other methods (most likely `FromFilename`).

Reply:
They are created by those `From*` functions.

A decoder op will do `FromHostMem(tensor_data, tensor_vol, tensor_source_info)`

A reader op will do `FromFilename(image_filename)`

Reply:
You mean regular reader will do `FromFilename`?

Reply:
I meant a future fused reader that produces decoded images, as we are planning to have (for partial reading purposes)